server/base/src/main/java/org/apache/accumulo/server/util/FindCompactionTmpFiles.java

```diff
@@ -0,0 +1,215 @@
+          Arrays.stream(files).forEach(fs -> matches.add(fs.getPath()));
+        } catch (IOException e) {
+          LOG.error("Error looking for tmp files in volume: {}", vol, e);
+        }
+        return null;
+      }));
+    }
+    svc.shutdown();
+
+    while (futures.size() > 0) {
+      UtilWaitThread.sleep(10_000);
+      Iterator<Future<Void>> iter = futures.iterator();
+      while (iter.hasNext()) {
+        Future<Void> future = iter.next();
+        if (future.isDone()) {
+          iter.remove();
+          try {
+            future.get();
+          } catch (InterruptedException | ExecutionException e) {
+            throw new RuntimeException("Error getting list of tmp files", e);
+          }
+        }
+      }
+    }
+    svc.awaitTermination(10, TimeUnit.MINUTES);
+    return matches;
+  }
+
+  public static class DeleteStats {
+    public int success = 0;
+    public int failure = 0;
+    public int error = 0;
+  }
+
+  public static DeleteStats deleteTempFiles(ServerContext context, List<Path> filesToDelete)
+      throws InterruptedException {
+
+    final ExecutorService delSvc = Executors.newFixedThreadPool(8);
+    final List<Future<Boolean>> futures = new ArrayList<>(filesToDelete.size());
+    final DeleteStats stats = new DeleteStats();
+
+    filesToDelete.forEach(p -> {
+      futures.add(delSvc.submit(() -> context.getVolumeManager().delete(p)));
+    });
+    delSvc.shutdown();
+
+    int expectedResponses = filesToDelete.size();
+    while (expectedResponses > 0) {
+      UtilWaitThread.sleep(10_000);
+      Iterator<Future<Boolean>> iter = futures.iterator();
+      while (iter.hasNext()) {
+        Future<Boolean> future = iter.next();
+        if (future.isDone()) {
+          expectedResponses--;
+          iter.remove();
+          try {
+            if (future.get()) {
+              stats.success++;
+            } else {
+              stats.failure++;
+            }
+          } catch (ExecutionException e) {
+            stats.error++;
+            LOG.error("Error deleting a compaction tmp file", e);
+          }
+        }
+      }
+      LOG.debug("Waiting on {} responses", expectedResponses);
+    }
+    delSvc.awaitTermination(10, TimeUnit.MINUTES);
+    return stats;
+  }
+
+  public static void main(String[] args) throws Exception {
+    Opts opts = new Opts();
+    opts.parseArgs(FindCompactionTmpFiles.class.getName(), args);
+    LOG.info("Deleting compaction tmp files: {}", opts.delete);
+    Span span = TraceUtil.startSpan(FindCompactionTmpFiles.class, "main");
+    try (Scope scope = span.makeCurrent()) {
+      ServerContext context = opts.getServerContext();
+      if (!allCompactorsDown(context)) {
+        LOG.warn("Compactor addresses found in ZooKeeper. Unable to run this utility.");
+      }
+
+      final List<Path> matches = findTempFiles(context);
+      LOG.info("Found the following compaction tmp files:");
+      matches.forEach(p -> LOG.info("{}", p));
+
```

Comment:
At this point we could scan the metadata table and get active ECIDs and remove paths from matches that have and active ECID.  Scanning for only tablets that have an ECID would be another good use case for #3933.  Would need to scan all three levels of metadata.

That would make this utility able to be run while compactors are running.  If that is the case then the compaction coordinator could possibly call this code periodically (like once a day by default).   The other code for deleting temp files could miss files if the coordinator dies a certain time or because of race conditions with split and merge that delete the tablet and create new tablet.  Periodically running this would still clean up things missed by the other code.

Reply:
Interesting thought. I didn't consider running this periodically so that things were eventually clean.

Reply:
For filtering, could just use a scanner directly and only fetch the ECID column (don't need any other tablet columns like prev row, so don't need to filter at the row level for this case).  Would still need to use ample to read root table metadata.

Reply:
One concern I have for running this in the CompactionCoordinator, which is running in the Manager, is memory usage in the case where there are a *lot* of tmp files that need to be cleaned up. The call to `FileSystem.globStatus` is going to return an unknown amount of `FileStatus` objects.

Reply:
I talked with @keith-turner and we agreed to not put this in the Manager for now, due to the memory issues for the globStatus results. I am going to change the utility though so that it does not expect all of the Compactors to be down, so that it can be run at any time.

Reply:
Modified utility in 6e75f70 to remove the constraint that Compactors are down, and removed set of current active external compactions from the compaction tmp files that exist on the filesystem, in 6e75f70.